bufferevent_ssl.c

```diff
@@ -268,20 +268,33 @@ do_read(struct bufferevent_ssl *bev_ssl, int n_to_read) {
 	if (n < 0)
 		return OP_ERR;
 
-	for (i=0; i<n; ++i) {
+	int len = 0;
+	for (i = 0; i < n;) {
 		if (bev_ssl->bev.read_suspended)
 			break;
 		bev_ssl->ssl_ops->clear_error();
-		r = bev_ssl->ssl_ops->read(bev_ssl->ssl, space[i].iov_base, space[i].iov_len);
-		if (r>0) {
+		r = bev_ssl->ssl_ops->read(
+			bev_ssl->ssl, space[i].iov_base + len, space[i].iov_len - len);
+		if (r > 0) {
 			result |= OP_MADE_PROGRESS;
 			if (bev_ssl->read_blocked_on_write)
 				if (clear_rbow(bev_ssl) < 0)
 					return OP_ERR | result;
-			++n_used;
-			space[i].iov_len = r;
 			bev_ssl->ssl_ops->decrement_buckets(bev_ssl);
+			len += r;
+			if (space[i].iov_len - len < 16384) {
```

Comment:
>Once SSL_read() only get max 16K bytes (one TLS record).

I guess this is because the size of TLS record has this size, but this is decrypted data, which can be any size

Reply:
Yes, but maximum a record only 16K
https://www.openssl.org/docs/man1.0.2/man3/SSL_read.html

My idea is after one SSL_read() if remain space is not enough 16K than use next space.
Or you think we should change to following?
```
			if (space[i].iov_len - len > 0) {
				continue;
			} else {
				space[i].iov_len = len;
				len = 0;
				++i;
			}
```

Reply:
>Or you think we should change to following?

Yes, let's try.
Can you also measure the performance impact of this?

Reply:
Without this PR
```
Performance counter stats for 'bin/https-client -url https://dragon3/smooth/4s/asset-1516/asset-1516.ism/QualityLevels(2000000)/Fragments(video=0) -ignore-cert':

             18,77 msec task-clock                #    0,003 CPUs utilized          
               119      context-switches          #    6,341 K/sec                  
                 1      cpu-migrations            #   53,289 /sec                   
               290      page-faults               #   15,454 K/sec                  
        20.000.071      cycles                    #    1,066 GHz                    
        18.285.468      instructions              #    0,91  insn per cycle        
```
With this PR
```
Performance counter stats for 'bin/https-client -url https://dragon3/smooth/4s/asset-1516/asset-1516.ism/QualityLevels(2000000)/Fragments(video=0) -ignore-cert':

             17,46 msec task-clock                #    0,002 CPUs utilized          
                88      context-switches          #    5,039 K/sec                  
                 1      cpu-migrations            #   57,266 /sec                   
               291      page-faults               #   16,664 K/sec                  
        18.911.167      cycles                    #    1,083 GHz                    
        17.729.885      instructions              #    0,94  insn per cycle
```

Reply:
I thought about something more the just running `perf` :)
I.e measure the throughput for example, for upstream/master, initial version of the patch (no partial reads, of less then 16K) and with partial reads (with `if (space[i].iov_len - len > 0)`)

Reply:
I try to use `nload` and measure max incoming.
- With origin/master, max income is 2.26 MBit/s.
- With initial version of the patch: max income is 2.44 Mbit/s
- With partial reads (latest version): same max income 2.44 Mbit/s

Is this throughput you mean? If not, please tell me more detail how to measure.

Reply:
Yes, it is indeed what I wanted! So performance improved by 8%, great! Can you also attach the methodology of the testing in the commit please?

Reply:
ok, will write it in the commit

Reply:
Thanks!

>Note: edit macros READ_DEFAULT, MAX_SINGLE_READ_DEFAULT to 32K and
>add the chunk callback to read out each piece of data.

- `MAX_SINGLE_READ_DEFAULT` can changed with API - `bufferevent_set_max_single_read`
- `READ_DEFAULT` - I don't see any reason for not using `bufferevent_get_read_max_(&bev->bev)` here instead, what do you think?

Reply:
Yes, I think current code does not affect if app use `bufferevent_set_max_single_read`.
I make another commit before `SSL_read` fix